File: codigo/p1.c
/*
Programa 1

Eleva ao quadrado os N elementos de um vetor e calcula a soma.

*/

#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {

	int n;	
	if (argc == 2) {
		n = atoi(argv[1]);
		printf("%d\n", n);
	}

	srand(time(NULL));
	int i;
	double vetor[n];
	for (i = 0; i < n; i++) {
		// rand() gera um inteiro entre 0 e RAND_MAX
		vetor[i] = ((double)rand()/(double)RAND_MAX);
	}


	double somaQuadrados(double v[], int tam) {
		
		double soma;
		int i;

		soma = 0.0;
		for (i = 0; i < tam; i++) {
			soma += (v[i] * v[i]);
		}

		return soma;
	}

	printf("%f\n", somaQuadrados(vetor, n));

	return 0;
	
}

File: codigo/p1.go
//
// Programa 1
//
// Eleva ao quadrado os N elementos de um vetor e calcula a soma.

package main

import (
  "flag"
  "fmt"
  "math/rand"
  "strconv"
  )
  
func soma_quadrados_thread(vetor []float64, c chan float64) {
  soma := 0.
  for _, e := range vetor {
    soma += e*e
  }
  c <- soma
}


func main() {

  flag.Parse()
  arg := flag.Arg(0)
  n, _ := strconv.Atoi(arg)

  var vetor = make([]float64, n)
  for i := range vetor {
    vetor[i] = rand.Float64()
  }
  
  
  c := make(chan float64)
  go soma_quadrados_thread(vetor[:len(vetor)/2], c)
  go soma_quadrados_thread(vetor[len(vetor)/2:], c)
  
  a, b := <-c, <-c
  
  fmt.Println(a+b)
}